llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp

```diff
@@ -1991,9 +2068,155 @@ void FunctionEncoder::handleMemoryPhiNode(MemoryPhi &mphi, int passID) {
+
+z3.If(
+  z3.And(z3.BVAddNoOverflow(a_bv, b_bv, True), z3.BVAddNoUnderflow(a_bv, b_bv)),
+  res_i1_bv == 0,
+  res_i1_bv == 1
+)
+
+res_i32_bv == a_bv + b_bv
+*/
+void FunctionEncoder::handleAddSubWithOverflowIntrinsic(IntrinsicInst &i) {
+
+  outs() << "[handleIntrinsicAddWithOverflow] add_with_overflow intrinsic "
+         << "\n";
+  auto BBAsstVecIter = BBAssertionsMap.find(currentBB);
+  Intrinsic::ID intrinsicInstID = i.getIntrinsicID();
+  Value *intrinsicInstValue = dyn_cast<Value>(&i);
+  Value *callOperand0 = i.getOperand(0);
+  Value *callOperand1 = i.getOperand(1);
+
+  auto callOperand0BV = BitVecHelper::getBitVecSingValType(callOperand0);
+  auto callOperand1BV = BitVecHelper::getBitVecSingValType(callOperand1);
+  outs() << "[handleIntrinsicAddWithOverflow]"
+         << "callOperand0: " << *callOperand0
+         << " (bv: " << callOperand0BV.to_string().c_str() << ")"
+         << "\n";
+  outs() << "[handleIntrinsicAddWithOverflow]"
+         << "callOperand1: " << *callOperand1
+         << " (bv: " << callOperand1BV.to_string().c_str() << ")"
+         << "\n";
+  StructType *instRetType = dyn_cast<StructType>(i.getType());
+  if (!instRetType && instRetType->getNumElements() != 2) {
+    throw std::runtime_error(
+        "[handleIntrinsicAddWithOverflow]"
+        "Unsupported type for add_with_overflow intrinsic."
+        "Only struct types like {i32, i1} or {i64, i1} are supported\n");
+  }
+  outs() << "[handleIntrinsicAddWithOverflow]"
+         << "instRetType: " << *instRetType << "\n";
+
+  ValueBVTreeMap *currentBBValueBVTreeMap;
+  if (this->BBValueBVTreeMap.find(currentBB) == this->BBValueBVTreeMap.end()) {
+    currentBBValueBVTreeMap = new ValueBVTreeMap();
+    BBValueBVTreeMap.insert({currentBB, currentBBValueBVTreeMap});
+  } else {
+    currentBBValueBVTreeMap = this->BBValueBVTreeMap.at(currentBB);
+  }
+
+  // the call instruction returns a struct of type {i32, i1},
+  // create a bitvector tree of the same structure.
+  BVTree *newTree = setupBVTreeForArg(intrinsicInstValue,
+                                      intrinsicInstValue->getName().str());
+  outs() << "[handleIntrinsicAddWithOverflow] "
+         << "newTree: " << newTree->toString() << "\n";
+  currentBBValueBVTreeMap->insert({intrinsicInstValue, newTree});
+
+  auto resBV = newTree->getSubTree(0)->bv;
+  auto overflowBV = newTree->getSubTree(1)->bv;
+  outs() << "[handleIntrinsicAddWithOverflow] "
+         << "resBV: " << resBV.get_sort().to_string().c_str() << "\n";
+  outs() << "[handleIntrinsicAddWithOverflow] "
+         << "overflowBV: " << overflowBV.get_sort().to_string().c_str() << "\n";
+
+  z3::expr resultExprNoOverflowOrUnderflow(ctx), resultExpr(ctx);
+  if (intrinsicInstID == Intrinsic::sadd_with_overflow) {
+    resultExprNoOverflowOrUnderflow =
+        z3::ite((z3::bvadd_no_overflow(callOperand0BV, callOperand1BV, true) &&
+                 z3::bvadd_no_underflow(callOperand0BV, callOperand1BV)),
```

Comment:
Regarding Z3's semantics for underflows: Yeah, very weird (and annoying for us). I've never seen underflows being defined like that :confused: